### PR TITLE
Bob clone issues

### PR DESCRIPTION
**Pull request by 06rajesh**
- bob clone with git outputs print
- bob clone directory with the same name issues on the parent directory
- add `explicit-protocol` flag on `bob workspace add` to set the URL explicitly to the provided protocol
- add tests for `explicit-protocol` flag

### CODE REVIEW

File: test/e2e/add/add_test.go [hunk 1]
```diff
@@ -19,19 +19,27 @@ var _ = Describe("Test bob add", func() {
...
 
 		It("adds local repos to bob", func() {
 			for _, child := range childs {
-				Expect(b.Add(fmt.Sprintf("file://%s", child))).NotTo(HaveOccurred())
+				Expect(b.Add(fmt.Sprintf("file://%s", child), false)).NotTo(HaveOccurred())
 			}
 		})
 
+		It("adds HTTPS repo to bob, with explicit protocol", func() {
+			Expect(b.Add("https://github.com/pkg/requests", true)).NotTo(HaveOccurred())
```
Comment (Equanox) [inline test/e2e/add/add_test.go:36]: Pleasae check that sshurl is empty.

File: test/e2e/add/add_test.go [hunk 2]
```diff
@@ -19,19 +19,27 @@ var _ = Describe("Test bob add", func() {
...
 
 		It("adds SSH repo to bob", func() {
-			Expect(b.Add("git@github.com:Benchkram/bob.git")).NotTo(HaveOccurred())
+			Expect(b.Add("git@github.com:Benchkram/bob.git", false)).NotTo(HaveOccurred())
 		})
 
 		It("adds local repos to bob", func() {
 			for _, child := range childs {
-				Expect(b.Add(fmt.Sprintf("file://%s", child))).NotTo(HaveOccurred())
+				Expect(b.Add(fmt.Sprintf("file://%s", child), false)).NotTo(HaveOccurred())
```
Comment (Equanox) [inline test/e2e/add/add_test.go:31]: Pleasae check that sshurl & httpsurl is correct empty.

File: cli/cmd_root.go [hunk 3]
```diff
@@ -29,6 +29,7 @@ func init() {
 	rootCmd.AddCommand(cleanCmd)
 
 	// workspace
+	cmdAdd.Flags().Bool("explicit-protocol", false, "Set to repository protocol explicit to https/ssh depending on the given url")
```
Comment (Equanox) [inline cli/cmd_root.go:32]: As there is just this one flag on the add cmd a simpler version would also do.
"explicit-protocol" => "plain" ?

Also
```suggestion
	cmdAdd.Flags().Bool("plain", false, "Don't infer contrary protocol url (git/https) from scheme")
```

File: bob/clone.go [hunk 4]
```diff
@@ -84,3 +117,107 @@ func (b *B) CloneRepo(repoURL string) (_ string, err error) {
...
+
+	if repo.SSHUrl == "" {
+		ignoressh = true
+	}
+
+	if repo.HTTPSUrl == "" {
+		ignorehttp = true
+	}
+
+	if !ignoressh {
```
Comment (Equanox) [inline bob/clone.go:142]:
```suggestion
	if repo.SSHUrl != "" {
```

File: bob/clone.go [hunk 5]
```diff
@@ -84,3 +117,107 @@ func (b *B) CloneRepo(repoURL string) (_ string, err error) {
...
+		if err != nil {
+			return nil, err
+		}
+		urls = append(urls, cloneURLItem{
+			url:      repoFromSSH.SSH.String(),
+			protocol: "ssh",
+		})
+	}
+
+	if !ignorehttp {
```
Comment (Equanox) [inline bob/clone.go:153]:
```suggestion
	if repo.HTTPSUrl {
```

File: bob/clone.go [hunk 6]
```diff
@@ -84,3 +117,107 @@ func (b *B) CloneRepo(repoURL string) (_ string, err error) {
...
+	if !success {
+		title = fmt.Sprint(repopath, "\t", aurora.Red("error"))
+	}
+	fmt.Fprint(buf, title)
+	fmt.Fprintln(buf)
+
+	return buf
+}
+
+// formatRepoNameForOutput returns formatted reponame for output.
```
Comment (Equanox) [inline bob/clone.go:206]:
```suggestion
// formatRepoNameForOutput returns sanitized reponame.
```

File: bob/clone.go [hunk 7]
```diff
@@ -84,3 +117,107 @@ func (b *B) CloneRepo(repoURL string) (_ string, err error) {
...
+//
+// Example: "." => "/", "second-level" => "second-level/"
+func formatRepoNameForOutput(reponame string) string {
```
Comment (Equanox) [inline bob/clone.go:209]:
```suggestion
func sanitizeReponame(reponame string) string {
```

File: cli/cmd_workspace_add.go [hunk 8]
```diff
@@ -16,13 +16,19 @@ var cmdAdd = &cobra.Command{
 	Args:  cobra.ExactArgs(1),
 	Long:  ``,
 	Run: func(cmd *cobra.Command, args []string) {
+		explicit, err := cmd.Flags().GetBool("explicit-protocol")
```
Comment (Equanox) [inline cli/cmd_workspace_add.go:19]: plain instead of explicit, everywhere

File: bob/add.go [hunk 9]
```diff
@@ -21,12 +33,26 @@ func (b *B) Add(rawurl string) (err error) {
 		}
 	}
 
+	httpsstr := repo.HTTPS.String()
+	sshstr := repo.SSH.String()
+	localstr := repo.Local
+
+	if explcitprotcl && checkIfHttp(rawurl) {
```
Comment (Equanox) [inline bob/add.go:40]:
```
if explicit {
  switch getScheme(rawurl) {
    case http:
      fallthrough
    case https:
    case git:
    case file:
    default:
        ... err... 
  }
}
```

**Review by Equanox — CHANGES_REQUESTED** [review]

File: test/e2e/add/add_test.go (hunk 1 above)
Comment (06rajesh) [reply test/e2e/add/add_test.go]: test added for empty rawurl

File: cli/cmd_root.go (hunk 3 above)
Comment (06rajesh) [reply cli/cmd_root.go]: changed to `plain`

File: bob/clone.go (hunk 4 above)
Comment (06rajesh) [reply bob/clone.go]: changed

File: bob/clone.go (hunk 5 above)
Comment (06rajesh) [reply bob/clone.go]: changed

File: cli/cmd_workspace_add.go (hunk 8 above)
Comment (06rajesh) [reply cli/cmd_workspace_add.go]: changed

**Review by Equanox — APPROVED** [review]

**Pull request merged**